### PR TITLE
Reset metrics before each task starts

### LINKED ISSUE #7558: `callback_metrics` should be reset when a new epoch start
## 🐛 Bug

Callback metrics should be destroyed whenever a new epoch/loop starts.

### To Reproduce

```python
def test_bug(tmpdir):
    import pytorch_lightning as pl
    from tests.helpers.boring_model import BoringModel

    class TestModel(BoringModel):

        def on_train_start(self) -> None:
            assert not self.trainer.callback_metrics

        def training_step(self, batch, batch_idx):
            output = self.layer(batch)
            loss = self.loss(batch, output)
            self.log('train_loss', loss)
            return loss

        def validation_step(self, batch, batch_idx):
            output = self.layer(batch)
            loss = self.loss(batch, output)
            self.log('val_loss', loss)

    model = TestModel()
    model.validation_epoch_end = None
    trainer = pl.Trainer(
        default_root_dir=tmpdir,
        max_epochs=3,
        num_sanity_val_steps=0,
        callbacks=[pl.callbacks.ModelCheckpoint(monitor='val_loss')]
    )

    trainer.validate(model)
    print(trainer.callback_metrics)

    model.validation_step = None
    trainer.fit(model)
```

### Expected behavior

The above case should throw an error because during `.fit` validation is disabled but since `trainer.validate` runs before it was disabled, the callback_metrics calculated during that loop stays and doesn't cause a `MisconfigurationException` during checkpointing. This is just one case where wrong metrics stored can cause some issues.

Another one can occur if `check_val_every_n_epoch>1` where if validation is not happening in every epoch, it still contains metrics from some old epoch which can cause some issues with `ModelCheckpoint/EarlyStopping/ReduceLROnPlateau`.

### Environment

 - PyTorch Version (e.g., 1.0): 1.8.1+cu101
 - OS (e.g., Linux): Linux
 - How you installed PyTorch (`conda`, `pip`, source): pip
 - Build command you used (if compiling from source):
 - Python version: 3.7
 - CUDA/cuDNN version: 10.1
 - GPU models and configuration: Tesla T4

### PR DESCRIPTION
**Pull request by rohitgr7**
## What does this PR do?

Fixes #7558

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [ ] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

**Review by tchaton — APPROVED** [review]
LGMT !

**Review by awaelchli — COMMENTED** [review]
Given the description of the PR and looking at the implementation, I think it's better these calls go directly into the trainer.run implementation, not the loops.

**Comment by rohitgr7** [comment]
> Given the description of the PR and looking at the implementation, I think it's better these calls go directly into the trainer.run implementation, not the loops.

good point, thanks :)

File: pytorch_lightning/trainer/connectors/logger_connector/logger_connector.py [hunk 1]
```diff
@@ -278,14 +278,20 @@ def should_reset_tensors(self, fx: str) -> bool:
...
+        if trainer_metrics:
             # reset metrics
-            self._progress_bar_metrics = {}
-            self._logged_metrics = {}
-            self._callback_metrics = {}
-        assert self.trainer._results is not None
-        self.trainer._results.reset(metrics=metrics)
+            self.reset_metrics()
+
+        if self.trainer.state.fn != TrainerFn.PREDICTING:
```
Comment (awaelchli) [inline pytorch_lightning/trainer/connectors/logger_connector/logger_connector.py:291]: can you explain this check?

Comment (carmocca) [reply pytorch_lightning/trainer/connectors/logger_connector/logger_connector.py:291]: I can, the prediction loops do not define `_results` (they are `None`)

But i'd change the code to:


```python
if self.trainer._results is not None:
    self.trainer._results.reset(metrics=metrics)
```

Instead of the predict check and the assertion

File: pytorch_lightning/trainer/connectors/logger_connector/logger_connector.py [hunk 2]
```diff
@@ -278,14 +278,20 @@ def should_reset_tensors(self, fx: str) -> bool:
             is_first_batch = bool(self._batch_idx) + self._split_idx == 0
         return is_different_fx and is_first_batch
 
-    def reset(self, metrics: Optional[bool] = None) -> None:
-        if self.trainer.sanity_checking:
+    def reset_metrics(self) -> None:
```
Comment (carmocca) [inline pytorch_lightning/trainer/connectors/logger_connector/logger_connector.py:281]: Do not extract this if it's not going to be called from the outside.

I think it'd be better having two functions, this one and `def reset_results(metrics)`

to avoid the `trainer_metrics` flag

File: pytorch_lightning/trainer/trainer.py [hunk 3]
```diff
@@ -989,6 +989,9 @@ def _run(self, model: "pl.LightningModule") -> Optional[Union[_EVALUATE_OUTPUT,
         # ----------------------------
         # TRAIN
         # ----------------------------
+        # reset logger connector
```
Comment (carmocca) [inline pytorch_lightning/trainer/trainer.py:992]: It is easier to put these reset calls directly in the trainer but IMO they should be managed by each loop, even if it's more verbose.

File: pytorch_lightning/trainer/connectors/logger_connector/logger_connector.py (hunk 1 above)
Comment (rohitgr7) [reply pytorch_lightning/trainer/connectors/logger_connector/logger_connector.py:291]: Yeah this is better, Will update this.. thanks @carmocca

File: pytorch_lightning/trainer/trainer.py (hunk 3 above)
Comment (rohitgr7) [reply pytorch_lightning/trainer/trainer.py:992]: @awaelchli suggested this. It was in the loop before. Any reason why they should be inside loop?

File: pytorch_lightning/trainer/connectors/logger_connector/logger_connector.py (hunk 2 above)
Comment (rohitgr7) [reply pytorch_lightning/trainer/connectors/logger_connector/logger_connector.py:281]: I think it can be called from outside, let's say someone wants to reset their metrics after every epoch, then they can just use this in one of their hooks. Right now it resets them at every run.

File: pytorch_lightning/trainer/trainer.py (hunk 3 above)
Comment (carmocca) [reply pytorch_lightning/trainer/trainer.py:992]: because the loops are the ones to control the flow of training, and they should be responsible for interacting with the logger connector, not the trainer.
Otherwise, the customization is limited.

For example, compare the existing usage of the `logger_connector` in the trainer vs the loops.

File: pytorch_lightning/trainer/connectors/logger_connector/logger_connector.py (hunk 2 above)
Comment (carmocca) [reply pytorch_lightning/trainer/connectors/logger_connector/logger_connector.py:281]: That's why I suggest keeping it public but removing the `trainer_metrics` flag

File: pytorch_lightning/trainer/trainer.py (hunk 3 above)
Comment (awaelchli) [reply pytorch_lightning/trainer/trainer.py:992]: well here is some advice 😅 
in most cases where @carmocca's arguments stand against @awaelchli's arguments, @carmocca wins. So when @awaelchli suggests changes better wait until @carmocca reviews before you make the changes 😄

**Comment by rohitgr7** [comment]
due to recent fault-tolerant changes, I had to move it back to the trainer. Since `logger_connector` is sort of handled by the trainer, this seems ok to me.

**Comment by rohitgr7** [comment]
> > due to recent fault-tolerant changes, I had to move it back to the trainer
> 
> Can you elaborate on what are the changes and why they force moving it?

Training state is being restored before dispatch so adding these resets inside loops will reset the resultcollection set by resume_from_checkpoint.

File: CHANGELOG.md [hunk 4]
```diff
@@ -352,6 +352,15 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 - Fixed `add_argparse_args` raising `TypeError` when args are typed as `typing.Generic` in Python 3.6 ([#9554](https://github.com/PyTorchLightning/pytorch-lightning/pull/9554))
 
 
+- Fixed error reporting in DDP process reconciliation when processes are launched by an external agent ([#9389](https://github.com/PyTorchLightning/pytorch-lightning/pull/9389))
```
Comment (awaelchli) [inline CHANGELOG.md:355]: don't forget to resolve the bad merge here :)

File: pytorch_lightning/loops/dataloader/prediction_loop.py [hunk 5]
```diff
@@ -78,7 +78,7 @@ def reset(self) -> None:
         self.epoch_batch_indices = []
 
     def on_run_start(self) -> None:
-        """Calls ``on_predict_start`` hook."""
+        """Calls ``_on_predict_start`` hook."""
```
Comment (awaelchli) [inline pytorch_lightning/loops/dataloader/prediction_loop.py:81]: here too :)

**Review by awaelchli — APPROVED** [review]

**Review by Borda — APPROVED** [review]
